gem/gem.py

```diff
@@ -782,6 +783,54 @@ def __new__(cls, i, j):
         return self
 
 
+class Inverse(Node):
+    """ Inverse is a shape to shape operation.
+    """
+    __slots__ = ('children', 'shape')
+    __back__ = ()
+
+    def __new__(cls, tensor):
+        assert len(tensor.shape) == 2
+        assert tensor.shape[0] == tensor.shape[1]
+
+        # Invert 1x1 matrix
+        if tensor.shape == (1, 1):
+            multiindex = (Index(extent=1), Index(extent=1))
+            return ComponentTensor(Division(one, Indexed(tensor, multiindex)), multiindex)
+
+        self = super(Inverse, cls).__new__(cls)
+        self.children = (tensor,)
+        self.shape = tensor.shape
+
+        return self
+
+
+class Solve(Node):
+    """ Solve is a shape to shape operation similar to Inverse.
+
+        :arg A: Tensor
+        :arg B: Right-hand side
+        :arg multiindex: Indices for result of solve
+    """
+    __slots__ = ('children', 'shape')
+    __back__ = ()
+
+    def __init__(self, A, B):
+
+        # Children must have shape
+        assert A.shape
+        assert B.shape
+
+        # Shape requirements
+        assert len(A.shape) == 2
+        assert A.shape[0] == A.shape[1]
+        assert A.shape[0] == B.shape[0]
+
+        self.children = (A, B)
+        self.shape = A.shape[1:] + B.shape[1:]
+        self.free_indices = A.free_indices + B.free_indices
```

Comment:
This is both unneeded and wrong. Wrong because shared free indices between A and B would be listed twice, and because free indices are not sorted to their canonical ordering.  It is not needed, because `NodeMeta` calculates automatically (and correctly).